### ISSUE TITLE
Source.keepAlive stops producing messages arbitrarily

### ISSUE THREAD

**Issue by vilunov** [akka/akka#28993]
This is tricky to preproduce locally – in fact, we couldn't do it.

We used `keepAlive` to maintain an SSE HTTP connection, because due to a bug in akka-http (https://github.com/akka/akka-http/issues/1977, https://github.com/akka/akka-http/issues/926), we can't track whether the client terminated connection until we send a next message in that connection. `keepAlive` in our case sends a ping each second if the original source is idle. We found out that about 0.365% of connections terminated by connection timeout anyway, i.e. by the proxying nginx read timeout setting. We solved this by replacing `keepAlive` with `merge` with `tick`:

```scala
val tick = Source.tick(
  initialDelay = sseEndpointConfig.pingInterval,  // 1 second
  interval = sseEndpointConfig.pingInterval,  // 1 second
  tick = ServerSentEvent("", `type` = Ping),
)
val src = sse
  .merge(tick, eagerComplete = true)
  .takeWithin(sseEndpointConfig.maxConnectionInterval)  // 5 minutes
```

Implementation with `keepAlive`:

```scala
val src = sse
  .keepAlive(sseEndpointConfig.pingInterval, () => ServerSentEvent("", `type` = Ping))
  .takeWithin(sseEndpointConfig.maxConnectionInterval)  // 5 minutes
```

Our original source `sse` does not terminate by itself (except in a case of a fatal exception), but it is limited in life time by `takeWithin` (after `keepAlive`); also, clients can disconnect freely.

Below is a screenshot of an idle network dump. As you can see, it was sending a message each second, until number 301, after which it was terminated after 1 minute timeout.

![Снимок экрана 2020-04-29 в 10 49 38](https://user-images.githubusercontent.com/16502164/80573536-cf450580-8a08-11ea-8d4e-50570dfd21a1.png)

**Comment by patriknw** [comment]
cc @raboof @jrudolph

**Comment by jrudolph** [comment]
I could see two reasons:

 1. the TCP stream is backpressured (but the tcpdump definitely doesn't look like that)
 2. for some reason the `IdleInject` stage "misses" the timeout. I.e. the `onTimer` method:

https://github.com/akka/akka/blob/0e4d41ad33dbeb00b598cb75b4d29899371bdc8c/akka-stream/src/main/scala/akka/stream/impl/Timers.scala#L266-L273

is invoked but the condition evaluates to false. IMO the condition should be true in all cases, otherwise it would be a bug like this, so it should really be an assertion. The only way that I can see would be that the timer triggers *too early* and therefore the deadline condition isn't met.

Or maybe it's this: `System.nanoTime()` doesn't have nanosecond granularity (for me consecutive invocations never show values smaller than 30) despite the name (it jumps between multiple calls). So, it could be it's because the rounded time is exactly the deadline and the comparison is `<` instead of `<=`.

Could you try copying the implementation of `IdleInject` and replace the `<` by `<=`?

**Comment by jrudolph** [comment]
> Or maybe it's this: `System.nanoTime()` doesn't have nanosecond granularity (for me consecutive invocations never show values smaller than 30) despite the name (it jumps between multiple calls). So, it could be it's because the rounded time is exactly the deadline and the comparison is `<` instead of `<=`.

I tried to reproduce this but usually the timer was run 1 or 2 *milliseconds* later than scheduled even when I moved down the scheduler tick duration to 1ms. In general, it seems the scheduler usually schedules tasks 1-2 ticks too late. Maybe there are some machines where `System.nanoTime` is even so far less accurate? Seems hard to believe.

**Comment by jrudolph** [comment]
> In general, it seems the scheduler usually schedules tasks 1-2 ticks too late.

That's documented for `LightArrayRevolverScheduler`.

**Comment by jrudolph** [comment]
Hopefully fixed by #29091.
